src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala

```diff
@@ -11,90 +11,109 @@ import org.broadinstitute.dsde.workbench.sam._
     } *> IO.pure(group)
   }
 
-  override def loadGroup(groupName: WorkbenchGroupName): IO[Option[BasicWorkbenchGroup]] = cs.evalOn(ecForLdapBlockingIO) {
-    IO(Option(ldapConnectionPool.getEntry(groupDn(groupName))) map(unmarshalGroup))
+  override def loadGroup(groupName: WorkbenchGroupName): IO[Option[BasicWorkbenchGroup]] = {
+    for {
+      searchResult <- executeLdap(IO(ldapConnectionPool.getEntry(groupDn(groupName))))
+      res <- Option(searchResult).traverse{
+        entry =>
+          unmarshalGroup(entry).fold(err => IO.raiseError(new WorkbenchException(err)), r => IO.pure(r))
+      }
+    } yield res
   }
 
-  private def unmarshalGroup(results: Entry): BasicWorkbenchGroup = {
-    val cn = getAttribute(results, Attr.cn).getOrElse(throw new WorkbenchException(s"${Attr.cn} attribute missing: ${results.getDN}"))
-    val email = getAttribute(results, Attr.email).getOrElse(throw new WorkbenchException(s"${Attr.email} attribute missing: ${results.getDN}"))
-    val memberDns = getAttributes(results, Attr.uniqueMember)
+  private def unmarshalGroup(results: Entry): Either[String, BasicWorkbenchGroup] =
+    for {
+      cn <- getAttribute(results, Attr.cn).toRight(s"${Attr.cn} attribute missing: ${results.getDN}")
+      email <- getAttribute(results, Attr.email).toRight(s"${Attr.email} attribute missing: ${results.getDN}")
+      memberDns = getAttributes(results, Attr.uniqueMember)
+    } yield BasicWorkbenchGroup(WorkbenchGroupName(cn), memberDns.map(dnToSubject), WorkbenchEmail(email))
 
-    BasicWorkbenchGroup(WorkbenchGroupName(cn), memberDns.map(dnToSubject), WorkbenchEmail(email))
-  }
+  private def unmarshalGroupThrow(results: Entry): BasicWorkbenchGroup = unmarshalGroup(results).fold(s => throw new WorkbenchException(s), identity)
 
-  override def loadGroups(groupNames: Set[WorkbenchGroupName]): Future[Seq[BasicWorkbenchGroup]] = Future {
+  override def loadGroups(groupNames: Set[WorkbenchGroupName]): IO[Stream[BasicWorkbenchGroup]] = {
     val filters = groupNames.grouped(batchSize).map(batch => Filter.createORFilter(batch.map(g => Filter.createEqualityFilter(Attr.cn, g.value)).asJava)).toSeq
-    ldapSearchStream(groupsOu, SearchScope.SUB, filters:_*)(unmarshalGroup)
+
+    executeLdap(IO(ldapSearchStream(groupsOu, SearchScope.SUB, filters: _*)(unmarshalGroupThrow)))
   }
 
-  override def loadGroupEmail(groupName: WorkbenchGroupName): Future[Option[WorkbenchEmail]] = Future {
-    Option(ldapConnectionPool.getEntry(groupDn(groupName), Attr.email)) map { results =>
-      WorkbenchEmail(getAttribute(results, Attr.email).getOrElse(throw new WorkbenchException(s"${Attr.email} attribute missing: ${results.getDN}")))
-    }
+  override def loadGroupEmail(groupName: WorkbenchGroupName): IO[Option[WorkbenchEmail]] = {
+    for{
+      searchResult <- executeLdap(IO(ldapConnectionPool.getEntry(groupDn(groupName), Attr.email)))
+      res <- Option(searchResult).traverse{
+        e =>
```

Comment:
better val naming please wrt res and e

Reply:
I also think this could just be
```
    executeLdap(IO(Option(ldapConnectionPool.getEntry(groupDn(groupName), Attr.email))) map { results =>
      WorkbenchEmail(getAttribute(results, Attr.email).getOrElse(throw new WorkbenchException(s"${Attr.email} attribute missing: ${results.getDN}")))
    }
```